features/mbedtls/targets/TARGET_STM/sha256_alt.c

```diff
@@ -0,0 +1,141 @@
+/*
+ *  sha256_alt.c for SHA256 HASH
+ *******************************************************************************
+ * Copyright (c) 2017, STMicroelectronics
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+#include "mbedtls/sha256.h"
+
+#if defined(MBEDTLS_SHA256_ALT)
+#include "mbedtls/platform.h"
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
+    volatile unsigned char *p = v; while( n-- ) *p++ = 0;
+}
+
+void mbedtls_sha256_init( mbedtls_sha256_context *ctx )
+{
+    mbedtls_zeroize( ctx, sizeof( mbedtls_sha256_context ) );
+
+    /* Enable HASH clock */
+    __HAL_RCC_HASH_CLK_ENABLE();
+}
+
+void mbedtls_sha256_free( mbedtls_sha256_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+    /* Force the HASH Periheral Clock Reset */
+    __HAL_RCC_HASH_FORCE_RESET();
+
+    /* Release the HASH Periheral Clock Reset */
+    __HAL_RCC_HASH_RELEASE_RESET();
+
+    mbedtls_zeroize( ctx, sizeof( mbedtls_sha256_context ) );
+}
+
+void mbedtls_sha256_clone( mbedtls_sha256_context *dst,
+                         const mbedtls_sha256_context *src )
+{
+    *dst = *src;
+}
+
+void mbedtls_sha256_starts( mbedtls_sha256_context *ctx, int is224 )
+{
+    /* HASH IP initialization */
+    if (HAL_HASH_DeInit(&ctx->hhash_sha256) == HAL_ERROR) {
+        // error found to be returned
+        return;
+    }
+
+    ctx->is224 = is224;
+    /* HASH Configuration */
+    ctx->hhash_sha256.Init.DataType = HASH_DATATYPE_8B;
+    if (HAL_HASH_Init(&ctx->hhash_sha256) == HAL_ERROR) {
+        // error found to be returned
+        return;
+    }
+}
+
+void mbedtls_sha256_process( mbedtls_sha256_context *ctx, const unsigned char data[MBEDTLS_SHA256_BLOCK_SIZE] )
+{
+    if (ctx->is224 == 0) {
+        HAL_HASHEx_SHA256_Accumulate(&ctx->hhash_sha256, (uint8_t *) data, MBEDTLS_SHA256_BLOCK_SIZE);
```

Comment:
@andresag01 Yes, we should. But we don't have a way to return an error with the current API. So this is to be improved when we (mbed TLS team) provide a better API.

There is one possible improvement: instead of returning an error, write random data to the hash result buffer. This ensures that if an error happens, then the resulting hash value is distinct from any valid hash. This trick was suggested by Marc Stevens in a somewhat similar scenario (when a hash calculation fails but has no way to return a failure status — in Marc Stevens's case, the failure was that the calculation detected a probable collision attempt).

I'm not sure if we should currently recommend this as a workaround.